tools/inference_engine/src/impl/model.cpp

```diff
@@ -209,13 +217,30 @@ class FusionVisitor : public GpuVisitor {
         std::cout << "visiting activation...\n";
 
         // check matmul + activation and fuse?
+        // rules:
+        // 1) the last activation in the chain of activations can perform the fusion (remove nodes if needed)
         auto& inputs = pn->get_inputs();
         if (inputs.size() == 1) {
-            auto possibleMatMul = dynamic_cast<GpuMatMul*>(inputs[0]);
-            if (possibleMatMul && possibleMatMul->get_outputs().size() == 1) {
-                std::cout << "possible matmul + activation fusion...\n";
-                possibleMatMul->fuse_with(pn);
-                to_delete_.insert(pn);
+            // is this the last activation in the chain?
+            auto nextActivation = pn->get_outputs().empty() ? nullptr : dynamic_cast<GpuActivation*>(pn->get_outputs()[0]);
+            if (nextActivation == nullptr) {
+                // walk through inputs and check if it's an activation, until we find some node to fuse with
+                std::vector<GpuActivation*> activations;
+                auto curr_node = pn;
+                GpuMatMul* prev_to_fuse_with = nullptr;
```

Comment:
We know that there will be more OPs supporting fusion with activation. Is there possibility to define type of the OPs as template arguments?